File: ComputationalGeometry/ApplicationOfTriangulation/myLibraries/util/geometry/elements/point/EventEdge.java
package myLibraries.util.geometry.elements.point;

/*
 * EventEdge.java
 *
 * Version:
 *     $1.0$
 *
 * Revisions:
 *     $1.0$
 */

import myLibraries.util.geometry.DCEL.MonotoneVertex;
import myLibraries.util.geometry.elements.UpdateCoordinatesShape;
import myLibraries.util.geometry.elements.line.Line;

/**
 * Data structure of Event edge for partitioning monotone subpolygon
 *
 * @author       Xiaoyu Tongyang, or call me sora for short
 */

public class EventEdge extends Event {
    private static int IDStatic = 0;
    // helper for split vertex, but maybe a merge vertex
    public MonotoneVertex vertex;
    // line
    public final UpdateCoordinatesShape shape;

    /**
     * Constructs to create an instance of EventEdge
     * */

    public EventEdge( double x, double y,
                     UpdateCoordinatesShape shape,
                      MonotoneVertex vertex ) {
        super( x, y, IDStatic++ );
        this.shape = shape;
        this.vertex = vertex;
    }

    public EventEdge( UpdateCoordinatesShape shape,
                      MonotoneVertex vertex ) {
        this( shape, vertex, IDStatic++ );
    }

    public EventEdge( UpdateCoordinatesShape shape,
                      MonotoneVertex vertex, int ID ) {
        super( vertex.x, vertex.y, ID );
        this.shape = shape;
        this.vertex = vertex;
    }

    /**
     * update Y And X with the shape in this EventEdge
     *
     * */

    @Override
    public void updateYAndX( Vector update ) {
        if ( ( ( Line ) shape ).outOfRangeX( update.x ) ) return;
        shape.updateYAndX(this, update, false );
    }

    /**
     * update Y with given X with the shape in this EventEdge
     * */

    @Override
    public void updateY( double x ) {
        if ( ( ( Line ) shape ).outOfRangeX( x ) ) return;
        shape.updateYAndX( this, x );
    }

    @Override
    public String toString() {
        return toStringNormal() + "(h:" + vertex + ")";
    }
}

File: web/myLibraries/util/geometry/elements/point/EventEdge.js
"use strict"

/*
 * EventEdge.java
 *
 * Version:
 *     $1.0$
 *
 * Revisions:
 *     $1.0$
 */

import Vector from "./Vector.js";
import CompareElement from "../../../CompareElement.js";

/**
 * Data structure of Event edge for partitioning monotone subpolygon
 *
 * The source code in java is from my own github:
 * @see <a href=https://github.com/fengkeyleaf/Algorithm/blob/main/ComputationalGeometry/ApplicationOfTriangulation/myLibraries/util/geometry/elements/point/EventEdge.java>EventEdge</a>
 *
 * @author       Xiaoyu Tongyang, or call me sora for short
 */

export default class EventEdge extends Vector {
    static IDStatic = 0;

    /**
     * Constructs to create an instance of EventEdge
     * */

    constructor( paras ) {
        super( CompareElement.chooseWhich( paras.x, null ),
            CompareElement.chooseWhich( paras.y, null ), EventEdge._IDStatic++ );
        // line
        this.shape = CompareElement.chooseWhich( paras.shape, null );
        // helper for split vertex, but maybe a merge vertex
        this.vertex = CompareElement.chooseWhich( paras.vertex, null );
    }

    /**
     * update Y And X with the shape in this EventEdge
     *
     * @param {Vector} update
     * */

    updateYAndX( update ) {
        if ( this.shape.outOfRangeX( update.x ) ) return;
        this.shape.updateYAndX( this, update, false );
    }

    /**
     * update Y with given X with the shape in this EventEdge
     *
     * @param {Number} x
     * */

    updateY( x ) {
        if ( this.shape.outOfRangeX( x ) ) return;
        this.shape.updateYAndXByX( this, x );
    }

    toString() {
        return super.toString() + "(h:" + this.vertex.valueOf() + ")";
    }
}
